### ISSUE TITLE
LogBuilder skip allocation when no targets

### ISSUE THREAD

**Issue by snakefoot** [NLog/NLog#3433]
Could be nice if LogBuilder actually skipped the allocation of itself and LogEventInfo, when calling Logger with Level where no targets.

NLog.Fluent.LoggerExtensions  could instead return a dummy LogBuilder that doesn't do anything.

Would require one to remove the `Level`-method on the LogBuilder (Prevent it from dynamically changing level after initial level-assignment).

**Comment by snakefoot** [comment]
Alternative one could consider dropping LogBuilder completely, and instead create new extension-methods for Logger-interface that returns a LogEventInfo, an then create extension-methods for LogEventInfo (Extension methods for LogEventInfo will handle that no LogEventInfo is returned if loglevel has no targets).

Ex.:

`logger.Info().Write("hello {0}", world);`

**Comment by 304NotModified** [comment]
With fluent APIs it's in my opinion always better to return on intermediate object. (Until last call). That's more flexible and the dependencies are also better. So last paragraph I guess?

Breaking change I think ;). We have to keep in my that we should win enough for those Breaking changes.

**Comment by snakefoot** [comment]
Seems that `WithProperty` is working out nicely, and makes easier to apply structured logging to existing applications. I saw a nice trick where one did the following:

```
logger.LogProperties(new { OrderId = 42, UserId = 666, RequestId = Guid.NewGuid }).Info("Order Created");
```

Where the fluent-api performs reflection of the expando-object (can also just be `IDictionary<string,object>`). And harvest the property-names and values. This would also allow one to create a reusable expando-object with common-properties, where one just fill out what is available at the location:

```
logger.LogProperties(new LogContext() { OrderId = 42, UserId = 666 }).Info("Order Deleted");
```

This removes some of the magic-string logic from message-templates, and ensures the property-names are using the same standard everywhere. 

Maybe have the following ILogger extension methods (That returns a `LogBuilder`-struct)

- LogProperties(object value)
- LogProperty(string name, object value)
- LogException(Exception exception)
- LogCallsite() // Caller-member-attributes (not sure if they should happen automatically

The `LogBuilder`-struct only allocates (and performs reflection) when finally calling Debug/Info/Error-method, where it checks if the LogLevel is enabled.

**Comment by snakefoot** [comment]
Resolved by #4325
